Mathlib/RingTheory/Radical.lean

```diff
@@ -110,120 +173,282 @@ theorem radical_dvd_self (a : M) : radical a ∣ a := by
 
 variable {E : Type*} [EuclideanDomain E] [NormalizationMonoid E] [UniqueFactorizationMonoid E]
+  {a b u x : E}
 
 /-- Division of an element by its radical in an Euclidean domain. -/
 def divRadical (a : E) : E := a / radical a
 
-theorem radical_mul_divRadical (a : E) : radical a * divRadical a = a := by
-  rw [divRadical, ← EuclideanDomain.mul_div_assoc _ (radical_dvd_self a),
-    mul_div_cancel_left₀ _ (radical_ne_zero a)]
+theorem radical_mul_divRadical : radical a * divRadical a = a := by
+  rw [divRadical, ← EuclideanDomain.mul_div_assoc _ radical_dvd_self,
+    mul_div_cancel_left₀ _ radical_ne_zero]
 
-theorem divRadical_mul_radical (a : E) : divRadical a * radical a = a := by
+theorem divRadical_mul_radical : divRadical a * radical a = a := by
   rw [mul_comm]
-  exact radical_mul_divRadical a
+  exact radical_mul_divRadical
 
-theorem divRadical_ne_zero {a : E} (ha : a ≠ 0) : divRadical a ≠ 0 := by
-  rw [← radical_mul_divRadical a] at ha
+theorem divRadical_ne_zero (ha : a ≠ 0) : divRadical a ≠ 0 := by
+  rw [← radical_mul_divRadical (a := a)] at ha
   exact right_ne_zero_of_mul ha
 
-theorem divRadical_isUnit {u : E} (hu : IsUnit u) : IsUnit (divRadical u) := by
+theorem divRadical_isUnit (hu : IsUnit u) : IsUnit (divRadical u) := by
   rwa [divRadical, radical_of_isUnit hu, EuclideanDomain.div_one]
 
-theorem eq_divRadical {a x : E} (h : radical a * x = a) : x = divRadical a := by
-  apply EuclideanDomain.eq_div_of_mul_eq_left (radical_ne_zero a)
+theorem eq_divRadical (h : radical a * x = a) : x = divRadical a := by
+  apply EuclideanDomain.eq_div_of_mul_eq_left radical_ne_zero
   rwa [mul_comm]
 
-theorem divRadical_mul {a b : E} (hab : IsCoprime a b) :
+theorem divRadical_mul (hab : IsCoprime a b) :
     divRadical (a * b) = divRadical a * divRadical b := by
   symm; apply eq_divRadical
-  rw [radical_mul hab]
+  rw [UniqueFactorizationMonoid.radical_mul hab.isRelPrime]
   rw [mul_mul_mul_comm, radical_mul_divRadical, radical_mul_divRadical]
 
-theorem divRadical_dvd_self (a : E) : divRadical a ∣ a := by
-  exact Dvd.intro (radical a) (divRadical_mul_radical a)
+theorem divRadical_dvd_self (a : E) : divRadical a ∣ a :=
+  ⟨radical a, divRadical_mul_radical.symm⟩
 
 theorem _root_.IsCoprime.divRadical {a b : E} (h : IsCoprime a b) :
     IsCoprime (divRadical a) (divRadical b) := by
-  rw [← radical_mul_divRadical a] at h
-  rw [← radical_mul_divRadical b] at h
+  rw [← radical_mul_divRadical (a := a)] at h
+  rw [← radical_mul_divRadical (a := b)] at h
   exact h.of_mul_left_right.of_mul_right_right
 
 end EuclideanDomain
+
+section Nat
+
+lemma UniqueFactorizationMonoid.primeFactors_eq_natPrimeFactors :
+    primeFactors = Nat.primeFactors := by
+  ext n : 1
+  rw [primeFactors, Nat.factors_eq, Nat.primeFactors]
+  -- this convert is necessary because of the different DecidableEq instances
+  convert List.toFinset_coe _
+
+namespace Nat
+
+theorem radical_le_self {n : ℕ} (hn : n ≠ 0) : radical n ≤ n :=
```

Comment:
How about the slightly stronger:
```suggestion
@[simp] theorem radical_le_self_iff {n : ℕ} : radical n ≤ n ↔ 0 < n := by
```

Reply:
I think `radical_le_self` is the one that'll be most useful in practice, so it seems to me like the best option is just to have both? And similarly for `two_le_radical`: might as well just have both versions!